fipy/matrices/sparseMatrix.py

```diff
@@ -119,8 +119,176 @@ def addAtDiagonal(self, vector):
+        return (numerix.vstack([IDs] * M) + numerix.indices((M,N))[0] * self.mesh.globalNumberOfCells).flatten()
+
+    def _cellIDsToLocalColIDs(self, IDs):
+        M = self.numberOfVariables
+        N = len(IDs)
+        return (numerix.vstack([IDs] * M) + numerix.indices((M,N))[0] * self.mesh.numberOfCells).flatten()
+
+    @property
+    def _globalNonOverlappingColIDs(self):
+        return self._cellIDsToGlobalColIDs(self.mesh._globalNonOverlappingCellIDs)
+
+    @property
+    def _globalCommonColIDs(self):
+        return list(range(0, self.numberOfVariables, self.mesh.globalNumberOfCells))
+
+    @property
+    def _globalOverlappingColIDs(self):
+        return self._cellIDsToGlobalColIDs(self.mesh._globalOverlappingCellIDs)
+
+    @property
+    def _localOverlappingColIDs(self):
+        return self._cellIDsToLocalColIDs(self.mesh._localOverlappingCellIDs)
+
+    @property
+    def _localNonOverlappingColIDs(self):
+        return self._cellIDsToLocalColIDs(self.mesh._localNonOverlappingCellIDs)
+
+    def _getStencil(self, id1, id2, overlapping=False):
+        raise NotImplementedError
+
+    def _globalVectorAndIDs(self, vector, id1, id2, overlapping=False):
+        """Transforms local overlapping values and coordinates to global
+
+        Parameters
+        ----------
+        vector : array_like
+            The overlapping values to insert.
+        id1 : array_like
+            The local overlapping row indices.
+        id2 : array_like
+            The local overlapping column indices.
+        overlapping : bool
+            Whether to return ghosted values or not (default False)
+
+        Returns
+        -------
+        tuple of ndarray
+            (vector, global row indices, global column indices)
+        """
+        id1, id2, mask = self._getStencil(id1, id2, overlapping)
+        vector = vector[mask]
+        return (vector, id1, id2)
+
+    @property
+    def _bodies(self):
+        if not hasattr(self, "_bodies_"):
+            self._bodies_ = numerix.in1d(self.mesh._globalOverlappingCellIDs,
+                                         self.mesh._globalNonOverlappingCellIDs)
+        return self._bodies_
+
+    @property
+    def _ghosts(self):
+        if not hasattr(self, "_ghosts_"):
+            self._ghosts_ = self.mesh._globalOverlappingCellIDs[~self._bodies]
+            self._ghosts_ = self._cellIDsToGlobalRowIDs(self._ghosts_)
+            self._ghosts_ = self.orderer(self._ghosts_)
+
+        return self._ghosts_
+
+class _RowMesh2Matrix(_Mesh2Matrix):
+    def _getStencil(self, id1, id2, overlapping=False):
+        id1 = self._globalOverlappingRowIDs[id1]
+
+        if overlapping:
+            mask = numerix.ones(id1.shape, dtype=bool)
+        else:
+            mask = numerix.in1d(id1, self._globalNonOverlappingRowIDs)
+
+        id1 = self.orderer(id1[mask])
+        id2 = numerix.asarray(id2)[mask]
+
+        return id1, id2, mask
+
+class _ColMesh2Matrix(_Mesh2Matrix):
```

Comment:
Is this the same function but with id1 and id2 reversed? Might be worth merging them

Reply:
The global overlapping and nonoverlapping sets need to change, too, but still a good idea